File: Plugins/Orimath.Basics.View/ViewModels/PointViewModel.cs
using System;
using Orimath.Core;
using Orimath.Plugins;
using Orimath.Controls;

namespace Orimath.Basics.View.ViewModels
{
    public class PointViewModel : NotifyPropertyChanged, IDisplayTargetViewModel
    {
        private readonly ILayerModel? _layer;

        public PointViewModel(ILayerModel? layer, Point point, IViewPointConverter pointConverter)
        {
            _layer = layer;
            Source = point;
            (X, Y) = pointConverter.ModelToView(point);
        }

        public Point Source { get; }
        public double X { get; }
        public double Y { get; }

        public (ILayerModel, DisplayTarget) GetTarget()
        {
            if (_layer is null) throw new InvalidOperationException();
            return (_layer, DisplayTarget.NewPoint(Source));
        }
        public override string ToString() => Source.ToString();
    }
}

File: Plugins/Orimath.Basics.View/ViewModels/PointViewModel.fs
namespace Orimath.Basics.View.ViewModels
open Orimath.Core
open Orimath.Plugins
open Orimath.Controls
open Orimath.Basics.View

type PointViewModel(point: Point, pointConverter: IViewPointConverter, ?layer: ILayerModel) =
    inherit NotifyPropertyChanged()
    let (ScreenPoint(x, y)) = pointConverter.ModelToView(point)
    member _.Source = point
    member _.X = x
    member _.Y = y
    
    override _.ToString() = point.ToString()

    interface IDisplayTargetViewModel with
        member _.GetTarget() =
            match layer with
            | Some(l) -> l, DisplayTarget.Point(point)
            | None -> invalidOp "Layer is not set"
